### PR TITLE
#170947660 Implement Facebook and Google login

### PR DESCRIPTION
**Pull request by T2Wil**
#### What does this PR do?
 This PR helps a user to login with social networks: Facebook and Google.
#### Description of Task to be completed?
- Create  barefoot-nomad  API on Google cloud platform.
- Create barefoot-nomad app on Facebook for developers platform.
- Install passport package as an authentication middleware
- Install passport-google-oauth20 package to authenticate with Google.
- Install passport-facebook package to authenticate with Facebook.
- Connect user to DB
- Write tests
- Write documentation

#### How should this be manually tested?
- run `git clone https://github.com/andela/knights-bn-backend.git` to clone the repo.
- run `cd knights-bn-backend` to go to the project directory.
- run `git checkout ft-social-login-170947660` to switch to this branch.
- run `npm install` to install dependencies.
- create a Postgres database locally.
- create a .env file, paste in below variables and replace xxxx with the correct values: (Check code-knights `slack channel` to find the correct values)
```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
GOOGLE_CLIENT_ID = 'XXXXX'
GOOGLE_CLIENT_SECRET = 'XXXXX'
GOOGLE_CALLBACK_URL = 'XXXXX'
FCBK_CLIENT_ID = 'XXXXX'
FCBK_CLIENT_SECRET = 'XXXXX'
FCBK_CALLBACK_URL = 'XXXXX'

```
- run `npm run createTables` to create the database tables
- run `npm start` to start the server.
- Open your browser and copy this URL `http://localhost:4000/api/v1/auth/login/socialLogin`
- you should see a page with two buttons `Login with Google` and `Login with Facebook` fill in your credentials and follow the steps.

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[link](https://www.pivotaltracker.com/story/show/170947660)

### 1. social login page
image
### 2. Google consent screen
image
### 3. Retrieved user info
image

### 1. Fill-in the credentials
image
### 2. Get retrieved user info
image

### CODE REVIEW

**Review by Niyonsengaeric — COMMENTED** [review]
I get this error when I try to login using Facebook!
Screen Shot 2020-02-13 at 11 35 22

File: src/mockData/facebookAuth.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+const user = {
+  id: '1985674401577712', method: 'facebook', firstName: 'Ishimwe', lastName: 'William', email: 'ishimwewilliam@yahoo.fr',
```
Comment (Niyonsengaeric) [inline src/mockData/facebookAuth.js]: put each key-value pair on a new line. apply it in all `mockData` files.

File: src/page.html [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
```
Comment (Niyonsengaeric) [inline src/page.html]: put in a separate folder `templates`.

File: src/tests/socialLogin.test.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import dotenv from 'dotenv';
+import app from '../app';
+
+dotenv.config();
+process.env.NODE_ENV = 'test';
```
Comment (Niyonsengaeric) [inline src/tests/socialLogin.test.js]: this should not be here!

**Comment by Niyonsengaeric** [comment]
git `checkout ft-config-oauth-170947660` does not work

**Comment by T2Wil** [comment]
> I get this error when I try to login using Facebook!
> Screen Shot 2020-02-13 at 11 35 22

Thanks for the comment @Niyonsengaeric ,
To test a Facebook app access is required! 
So, I just created a test user that we might use  for testing purpose: 


image

Please use the following credentials on Facebook login
`email` : `bare-foot_ogwlfsl_nomad@tfbnw.net`
`password`: `Barefoot123`

File: src/mockData/facebookAuth.js (hunk 1 above)
Comment (T2Wil) [reply src/mockData/facebookAuth.js]: done, thanks!

File: src/page.html (hunk 2 above)
Comment (T2Wil) [reply src/page.html]: done!

File: src/tests/socialLogin.test.js (hunk 3 above)
Comment (T2Wil) [reply src/tests/socialLogin.test.js]: done!

**Comment by T2Wil** [comment]
> git `checkout ft-config-oauth-170947660` does not work

corrected, thanks @Niyonsengaeric

**Comment by T2Wil** [comment]
> git `checkout ft-config-oauth-170947660` does not work

thanks @Niyonsengaeric , I made the correction,

File: src/middlewares/getUserFromResourceServer.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+
+const getUser = (accessToken, refreshToken, profile, done) => {
+  console.log(`user : ${JSON.stringify(profile)}`);
```
Comment (MuhanguziDavid) [inline src/middlewares/getUserFromResourceServer.js]: Do we need this console statement?

File: src/middlewares/getUserFromResourceServer.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  const lastName = profile.name.lastName || profile.name.familyName;
+  const { gender } = profile;
+  const email = (profile.emails) ? profile.emails[0].value : null;
+  const user = {
+    id, method, firstName, lastName, email, gender,
+  };
+  done(null, user);
+};
+
+export default getUser;
```
Comment (MuhanguziDavid) [inline src/middlewares/getUserFromResourceServer.js]: This file doesnt seem to be tested! Run `npm run coverage` and look at the coverage report.

File: src/middlewares/getUserFromResourceServer.js (hunk 4 above)
Comment (T2Wil) [reply src/middlewares/getUserFromResourceServer.js]: Thanks @MuhanguziDavid , 
we definitely do not need this! 
I have deleted the line and moved the whole function into [passport config file](https://github.com/andela/knights-bn-backend/pull/27/files#diff-081b3f12542f954a81ba72aa794ddf46)

File: src/middlewares/getUserFromResourceServer.js (hunk 5 above)
Comment (T2Wil) [reply src/middlewares/getUserFromResourceServer.js]: I faced difficulties testing it but now is fixed. 
Thanks @MuhanguziDavid.

**Review by MuhanguziDavid — APPROVED** [review]

**Pull request merged**